### ISSUE TITLE
tfrecords generation uses incorrect data types?

### ISSUE THREAD

**Issue by fracpete** [waikato-ufdl/wai-annotations#30]
When turning **bboxes** (UFDL annotations) into TFRecords using:
```
to-tfrecords -o data.tfrecords -p labels.pbtxt --split-names train test val --split-ratios 70 15 15
````

I get the following error:
```
2020-10-12 20:03:01.652629: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at example_parsing_ops.cc:240 : Invalid argument: Key: image/object/area.  Data types don't match. Expected type: float, Actual type: int64
```

When converting **masks** using:
```
to-tfrecords -o data.tfrecords -p labels.pbtxt --split-names train test val --split-ratios 70 15 15
```

I get this error:

```
2020-10-12 20:06:20.770634: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at example_parsin
g_ops.cc:240 : Invalid argument: Key: image/object/mask.  Data types don't match. Expected type: float, Actual type: str
ing
```

**Comment by csterling** [comment]
@fracpete What format are you converting from?

**Comment by fracpete** [comment]
From the UFDL backend.

**Comment by csterling** [comment]
@fracpete Thanks. And also, does `wai.annotations` itself produce this error during conversion, or is it something else that tries to consume the `.tfrecords` files produced that has the error?

**Comment by fracpete** [comment]
The tfrecords generation with wai.annotations succeeds. It is Tensorflow/ObjectDetection API that then produces the error messages listed in the first post. It seems that the tfrecords generation has changed and can no longer be used by TF/ObjDet.

**Comment by csterling** [comment]
@fracpete Hmmm damn if it was occurring during conversion that would have meant there is some sort of canonical definition of these keys... Are there negative examples in your dataset?

**Comment by fracpete** [comment]
Negative examples as in explicit background images? No.

**Comment by csterling** [comment]
@fracpete Okay, then there is only one place in the conversion code where the error can be occurring by my estimation, and I can see a few different ways it might be failing, if the input data to that point were erroneous in some way. But of those few ways, the odds of them failing in the way they seem to be failing, without failing prior to that, seem astronomical.

To test this further, I'm going to need the dataset and the full `wai.annotations` command you are using, please.

**Comment by csterling** [comment]
@fracpete Okay I've fixed the issue with the bbox conversion. The areas being output were integers (the area of a box with integer side-lengths is an integer). These are now converted to floats as expected.

However the issue with the masks is more elusive. The `image/object/mask` field is, as I understand it, a bit-mask of the image with pixels covered by the annotation. Hence its type is a string (actually a binary string containing an entire PNG image encoding the mask), but the error above says it should be a float. So this falls into the category of problems, 'where are these TFRecords formats defined?'.

**Comment by csterling** [comment]
[This document](https://github.com/floydhub/object-detection-template/blob/master/object_detection/g3doc/instance_segmentation.md) outlines two types of mask formats, one which is the PNG encoding specified above, and another which is a "dense numerical" format which uses floats as the error suggests.

**Comment by fracpete** [comment]
Tensorflow Object Detection API v1.13.0 has a tool for converting COCO to TFRecords:
https://github.com/tensorflow/models/blob/v1.13.0/research/object_detection/dataset_tools/create_coco_tf_record.py
The area is a float list:
```python
      'image/object/area':
          dataset_util.float_list_feature(area),
```

**Comment by csterling** [comment]
That is the code that parts of the TF format conversion code is based on. It also has the capability to handle masks, but also does so using an encoded PNG image...

**Comment by csterling** [comment]
@fracpete According to [this issue](https://github.com/tensorflow/models/issues/5582), the `.config` file needs a field `mask_type: PNG_MASKS` to work. Is this enabled in your setup?

**Comment by fracpete** [comment]
No, only `load_instance_masks: true`. Are you currently storing the masks as PNGs?

**Comment by csterling** [comment]
Yeah, it looks like there are two possible formats, but `wai.annotations` has always used a PNG encoding.

**Comment by csterling** [comment]
It looks like the `mask_type: PNG_MASKS` should go directly after `load_instance_masks: true` in the config. I will create an issue for adding an option to `wai.annotations` to choose the mask encoding.

**Comment by csterling** [comment]
The fix for the bounding-box issue is implemented in 7dd629bdd63221f53b253a09fc4550952e0a1e13.

Either #31 or the use of `mask_type: PNG_MASKS` can be used to resolve the masks issue.

**Comment by csterling** [comment]
Released in [v0.5.1](https://github.com/waikato-ufdl/wai-annotations/releases/tag/v0.5.1)
